### PR TITLE
rtlwifi_new: fix build & rename to rtw88

### PR DESCRIPTION
**Pull request by samuelgrf**
###### Motivation for this change
The repository was moved to https://github.com/lwfinger/rtw88.
Tested with RTL8821CE chip.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuelgrf** [comment]
Result of `nixpkgs-review pr 116292` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxPackages-libre.rtlwifi_new</li>
    <li>linuxPackages_latest-libre.rtlwifi_new</li>
  </ul>
</details>
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>linuxPackages.rtlwifi_new</li>
    <li>linuxPackages_4_14.rtlwifi_new</li>
    <li>linuxPackages_4_19.rtlwifi_new</li>
    <li>linuxPackages_5_10.rtlwifi_new</li>
    <li>linuxPackages_5_11.rtlwifi_new</li>
    <li>linuxPackages_hardened.rtlwifi_new</li>
    <li>linuxPackages_latest_hardened.rtlwifi_new</li>
    <li>linuxPackages_latest_xen_dom0.rtlwifi_new</li>
    <li>linuxPackages_latest_xen_dom0_hardened.rtlwifi_new</li>
    <li>linuxPackages_lqx.rtlwifi_new</li>
    <li>linuxPackages_testing_bcachefs.rtlwifi_new</li>
    <li>linuxPackages_xen_dom0.rtlwifi_new</li>
    <li>linuxPackages_xen_dom0_hardened.rtlwifi_new</li>
    <li>linuxPackages_zen.rtlwifi_new</li>
    <li>rtlwifi_new-firmware</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 116292` at 602d6d8e run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxPackages-libre.rtlwifi_new</li>
    <li>linuxPackages-libre.rtw88</li>
    <li>linuxPackages_hardkernel_4_14.rtlwifi_new</li>
    <li>linuxPackages_hardkernel_4_14.rtw88</li>
    <li>linuxPackages_hardkernel_latest.rtlwifi_new</li>
    <li>linuxPackages_hardkernel_latest.rtw88</li>
    <li>linuxPackages_latest-libre.rtlwifi_new</li>
    <li>linuxPackages_latest-libre.rtw88</li>
    <li>linuxPackages_testing_bcachefs.rtlwifi_new</li>
    <li>linuxPackages_testing_bcachefs.rtw88</li>
  </ul>
</details>
<details>
  <summary>10 packages built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/5413d949cb4e25ecd54a1476e2c78a91">linuxPackages.rtw88</a></li> (linuxPackages.rtlwifi_new ,linuxPackages_5_4.rtlwifi_new ,linuxPackages_5_4.rtw88 ,linuxPackages_xen_dom0.rtlwifi_new ,linuxPackages_xen_dom0.rtw88)</li>
    <li><a href="https://gist.github.com/c777c6c6853a47f3d6015ac92e3d4f59">linuxPackages_4_14.rtw88</a></li> (linuxPackages_4_14.rtlwifi_new)</li>
    <li><a href="https://gist.github.com/9d217b337c23bfe564c6272acd14a492">linuxPackages_4_19.rtw88</a></li> (linuxPackages_4_19.rtlwifi_new)</li>
    <li><a href="https://gist.github.com/b5ed230dac14196e8d0fee6ec8b11c3d">linuxPackages_5_10.rtw88</a></li> (linuxPackages_5_10.rtlwifi_new)</li>
    <li><a href="https://gist.github.com/7d0ff88f9084592ee046424cccb914ab">linuxPackages_5_11.rtw88</a></li> (linuxPackages_5_11.rtlwifi_new ,linuxPackages_latest.rtlwifi_new ,linuxPackages_latest.rtw88 ,linuxPackages_latest_xen_dom0.rtlwifi_new ,linuxPackages_latest_xen_dom0.rtw88)</li>
    <li><a href="https://gist.github.com/29c016b46753c605a258c088b84bc768">linuxPackages_hardened.rtw88</a></li> (linuxPackages_hardened.rtlwifi_new ,linuxPackages_xen_dom0_hardened.rtlwifi_new ,linuxPackages_xen_dom0_hardened.rtw88)</li>
    <li><a href="https://gist.github.com/8d3bb67a0e87cfc23dc6124decfb5e9d">linuxPackages_latest_hardened.rtw88</a></li> (linuxPackages_latest_hardened.rtlwifi_new ,linuxPackages_latest_xen_dom0_hardened.rtlwifi_new ,linuxPackages_latest_xen_dom0_hardened.rtw88)</li>
    <li><a href="https://gist.github.com/c2ff865676f1790fe2dbdf6a891ff2b3">linuxPackages_lqx.rtw88</a></li> (linuxPackages_lqx.rtlwifi_new)</li>
    <li><a href="https://gist.github.com/61a629c007a9bf5c2bd8731fea386bb2">linuxPackages_zen.rtw88</a></li> (linuxPackages_zen.rtlwifi_new)</li>
    <li><a href="https://gist.github.com/e54691ddd76e530f6a4fcb6a022a7d89">rtw88-firmware</a></li> (rtlwifi_new-firmware)</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/os-specific/linux/rtw88/default.nix:34:5:
```
   |
34 |     license = licenses.gpl2;
   |     ^
```
</li>
  </ul>
</details>

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -20183,7 +20184,8 @@ in
 
   rtl8761b-firmware = callPackage ../os-specific/linux/firmware/rtl8761b-firmware { };
 
-  rtlwifi_new-firmware = callPackage ../os-specific/linux/firmware/rtlwifi_new-firmware { };
+  rtw88-firmware = callPackage ../os-specific/linux/firmware/rtw88-firmware { };
+  rtlwifi_new-firmware = rtw88-firmware;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20188]: same as above.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -19601,7 +19601,8 @@ in
 
     rtl8821cu = callPackage ../os-specific/linux/rtl8821cu { };
 
-    rtlwifi_new = callPackage ../os-specific/linux/rtlwifi_new { };
+    rtw88 = if lib.versionAtLeast kernel.version "4.14" then callPackage ../os-specific/linux/rtw88 { } else null;
+    rtlwifi_new = rtw88;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:19605]: This should be moved to aliases.nix with a throw.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (samuelgrf) [reply pkgs/top-level/all-packages.nix]: Moving this to aliases.nix doesn't work as it is a part of the `linuxPackagesFor` function, so I changed the already existing line to a throw.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -19601,7 +19601,8 @@ in
 
     rtl8821cu = callPackage ../os-specific/linux/rtl8821cu { };
 
-    rtlwifi_new = callPackage ../os-specific/linux/rtlwifi_new { };
+    rtw88 = if lib.versionAtLeast kernel.version "4.14" then callPackage ../os-specific/linux/rtw88 { } else null;
+    rtlwifi_new = throw "rtlwifi_new has been renamed to rtw88"; # added 2021-03-14
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:19605]:
```suggestion
```

File: pkgs/top-level/aliases.nix [hunk 4]
```diff
@@ -579,6 +579,7 @@ mapAliases ({
   rocm-runtime-ext = throw "rocm-runtime-ext has been removed, since its functionality was added to rocm-runtime"; #added 2020-08-21
   rssglx = rss-glx; #added 2015-03-25
   rssh = throw "rssh has been removed from nixpkgs: no upstream releases since 2012, several known CVEs"; # added 2020-08-25
+  rtlwifi_new-firmware = throw "rtlwifi_new-firmware has been renamed to rtw88-firmware"; # added 2021-03-14
```
Comment (SuperSandro2000) [inline pkgs/top-level/aliases.nix:582]:
```suggestion
  rtw88 = rtlwifi_new; # added 2021-03-14
  rtw88-firmware = rtlwifi_new-firmware; # added 2021-03-14
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by samuelgrf** [comment]
@SuperSandro2000 Applying the suggestions makes `nix build -f linuxPackages_latest.rtlwifi_new` fail with: `attribute [...] not found`.

Kernel modules are defined inside the `linuxPackagesFor` function, there's a comment explaining this: https://github.com/NixOS/nixpkgs/blob/e3e59d4ab7b9e353d1762100dfeafff66a4118ee/pkgs/top-level/all-packages.nix#L19489-L19493
I've also added a comment to aliases.nix explaining why the alias is in `all-packages.nix`.

**Review by TvoroG — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -19601,7 +19601,8 @@ in
 
     rtl8821cu = callPackage ../os-specific/linux/rtl8821cu { };
 
-    rtlwifi_new = callPackage ../os-specific/linux/rtlwifi_new { };
+    rtw88 = if lib.versionAtLeast kernel.version "4.14" then callPackage ../os-specific/linux/rtw88 { } else null;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:19604]: Changing a package to null is not allowed. Please set disabled, broken or similar.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (samuelgrf) [reply pkgs/top-level/all-packages.nix]: Done!

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -19601,7 +19601,11 @@ in
 
     rtl8821cu = callPackage ../os-specific/linux/rtl8821cu { };
 
-    rtlwifi_new = callPackage ../os-specific/linux/rtlwifi_new { };
+    rtw88 = if lib.versionAtLeast kernel.version "4.14"
+    then callPackage ../os-specific/linux/rtw88 { }
+    else (callPackage ../os-specific/linux/rtw88 { }).overrideAttrs (_: { meta.broken = true; });
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:19606]: Why are you not setting meta broken in the package itself? An example https://github.com/NixOS/nixpkgs/blob/8cb1afae1b57daeeb248ce54534398ccb9fd4e2d/pkgs/os-specific/linux/ena/default.nix#L44

Comment (samuelgrf) [reply pkgs/top-level/all-packages.nix]: Thats much cleaner, I don't know why I didn't think of that.
Thanks for taking the time to review this btw!

**Comment by samuelgrf** [comment]
The last force-push updates the package to the most recent revision.
Built and tested to be working with the RTL8821CE BT/WLAN chip.

File: pkgs/os-specific/linux/rtw88/default.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    find . -name '*.ko' -exec cp --parents {} ${modDestDir} \;
+    find ${modDestDir} -name '*.ko' -exec xz -f {} \;
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "The newest Realtek rtlwifi codes";
+    homepage = "https://github.com/lwfinger/rtw88";
+    license = licenses.gpl2;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/rtw88/default.nix:34]: gpl2Only or gpl2Plus?

File: pkgs/os-specific/linux/firmware/rtw88-firmware/default.nix [hunk 8]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, lib, linuxPackages }:
+
+stdenv.mkDerivation {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/firmware/rtw88-firmware/default.nix:3]:
```suggestion
stdenvNoCC.mkDerivation {
```

File: pkgs/os-specific/linux/firmware/rtw88-firmware/default.nix [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, lib, linuxPackages }:
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/firmware/rtw88-firmware/default.nix:1]:
```suggestion
{ stdenvNoCC, lib, linuxPackages }:
```

File: pkgs/os-specific/linux/rtw88/default.nix (hunk 7 above)
Comment (samuelgrf) [reply pkgs/os-specific/linux/rtw88/default.nix]: It's a dual licensed: BSD 3 & GPL 2.0 only

source: https://github.com/lwfinger/rtw88/blob/d4fd1ae72e306a9ca0a873c25937b639da54158b/rtw8821ce.c#L1

**Comment by samuelgrf** [comment]
@SuperSandro2000 I've applied the suggested changes, is there anything else that needs to be done before this can be merged?

**Comment by dasJ** [comment]
This broke eval: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/hardware/all-firmware.nix#L52

**Comment by samuelgrf** [comment]
@dasJ I've opened a PR fixing this: https://github.com/NixOS/nixpkgs/pull/118129.

**Pull request merged**